contracts/nexus-gateway/src/contract/execute.rs

```diff
@@ -86,23 +26,66 @@ pub fn route_to_router(
     Ok(Response::new().add_messages(router.route(msgs)))
 }
 
-pub fn route_to_nexus(
+pub fn route_message_with_token_to_nexus(
     storage: &mut dyn Storage,
+    querier: QuerierWrapper,
+    info: MessageInfo,
+    msg: router_api::Message,
+) -> Result<Response<nexus::execute::Message>> {
+    let token = match info.funds.as_slice() {
+        [token] => token.clone(),
+        _ => bail!(Error::InvalidToken(info.funds)),
+    };
+
+    route_message_to_nexus(storage, querier, msg, Some(token))?
+        .then(|msgs| Response::new().add_messages(msgs))
+        .then(Result::Ok)
+}
+
+pub fn route_messages_to_nexus(
+    storage: &mut dyn Storage,
+    querier: QuerierWrapper,
     msgs: Vec<router_api::Message>,
 ) -> Result<Response<nexus::execute::Message>> {
-    let msgs = msgs
-        .into_iter()
-        .filter_map(|msg| match state::is_message_routed(storage, &msg.cc_id) {
-            Ok(true) => None,
-            Ok(false) => Some(Ok(msg)),
-            Err(err) => Some(Err(err)),
-        })
-        .collect::<Result<Vec<_>>>()?;
+    msgs.into_iter()
+        .map(|msg| route_message_to_nexus(storage, querier, msg, None))
+        .collect::<Result<Vec<_>>>()?
+        .then(|msgs| msgs.concat())
+        .then(|msgs| Response::new().add_messages(msgs))
+        .then(Result::Ok)
+}
+
+fn route_message_to_nexus(
+    storage: &mut dyn Storage,
+    querier: QuerierWrapper,
+    msg: router_api::Message,
+    token: Option<Coin>,
+) -> Result<Vec<CosmosMsg<nexus::execute::Message>>> {
+    if state::is_message_routed(storage, &msg.cc_id)? {
+        return Ok(vec![]);
+    }
 
-    msgs.iter()
-        .try_for_each(|msg| state::set_message_routed(storage, &msg.cc_id))?;
+    state::set_message_routed(storage, &msg.cc_id)?;
 
-    let msgs: Vec<nexus::execute::Message> = msgs.into_iter().map(Into::into).collect();
+    let client: nexus::Client = client::CosmosClient::new(querier).into();
+    let config = load_config(storage)?;
+
+    let mut msg: nexus::execute::Message = msg.into();
+    msg.token.clone_from(&token);
 
-    Ok(Response::new().add_messages(msgs))
+    vec![client.route_message(msg)]
+        .then(move |mut msgs| {
+            if let Some(token) = token {
+                msgs.extend(iter::once(
+                    BankMsg::Send {
```

Comment:
The send will happen after the execute, but it should be done before so nexus has the token to process.

Reply:
good catch